src/main/cues/cue-emitter.js

```diff
@@ -0,0 +1,224 @@
+], function(EventEmitter, CueMark) {
+  'use strict';
+
+  var isHeadless = /HeadlessChrome/.test(navigator.userAgent);
+
+  function windowIsVisible() {
+    if (isHeadless || navigator.webdriver) {
+      return false;
+    }
+    return (typeof document === 'object') &&
+      ('visibilityState' in document) &&
+      (document.visibilityState === 'visible');
+  }
+
+  function getRAF(top) {
+    return top ? (
+      top.requestAnimationFrame ||
+      top.webkitRequestAnimationFrame ||
+      top.mozRequestAnimationFrame ||
+      top.msRequestAnimationFrame ||
+      function() {}
+    ) : function() {};
+  }
+  var rAF = getRAF(window);
+
+  /**
+   * This adapter populates the list of timeline entries from Peaks' points and segments
+   * NB: mutates marks, removing the old entries
+   * @param {CueMarkDef[]} marks
+   * @param {Peaks} peaks
+   */
+  function populateMarks(marks, peaks) {
+    var segments = peaks.segments.getSegments();
+    var points = peaks.points.getPoints();
+
+    marks.length = 0;
+    points.forEach(function(entry) {
+      marks.push(CueMark.create(entry.time, CueMark.POINT, entry.id));
+    });
+    segments.forEach(function(entry) {
+      marks.push(CueMark.create(entry.startTime, CueMark.SEGMENT_START, entry.id));
+      marks.push(CueMark.create(entry.endTime, CueMark.SEGMENT_END, entry.id));
+    });
+
+    marks.sort(CueMark.sorter);
+  }
+
+  /**
+   * @param {Peaks} peaks
+   * @return {HTMLMediaElement}
+   */
+  function getMediaElement(peaks) {
+    return peaks.player._mediaElement;
+  }
+
+  /**
+   * @class
+   * @property {Peaks} peaks
+   * @property {Array.<CueMarkDef>} marks
+   * @param peaks
+   * @constructor
+   */
+  function CueEmitter(peaks) {
+    EventEmitter.call(this);
+    this.marks = Array();
+    this.peaks = peaks;
+    this.previousTime = -1;
+    // bound to all Peaks events relating to mutated segments or points
+    this._updateMarks = this._updateMarks.bind(this);
+    // event handlers
+    this.onPeaksDestroyed = this.onPeaksDestroyed.bind(this);
+    this.onPlay = this.onPlay.bind(this);
+    this.onSeek = this.onSeek.bind(this);
+    this.onTimeUpdate = this.onTimeUpdate.bind(this);
+    this.onAnimationFrame = this.onAnimationFrame.bind(this);
+    this.init();
+  }
+
+  CueEmitter.prototype = Object.create(EventEmitter.prototype);
+  CueEmitter.prototype.constructor = CueEmitter;
+
+  CueEmitter.prototype._updateMarks = function() {
+    populateMarks(this.marks, this.peaks);
+  };
+
+  CueEmitter.prototype._onUpdate = function(time, previousTime) {
+    var marks = this.marks;
+    var isForward = time > previousTime;
+    var start;
+    var end;
+    var step;
+
+    if (isForward) {
+      start = 0;
+      end = marks.length;
+      step = 1;
+    }
+    else {
+      start = marks.length - 1;
+      end = -1;
+      step = -1;
+    }
+    // marks are sorted
+    for (var i = start, mark, markTime; isForward ? i < end : i > end; i += step) {
```

Comment:
As you mentioned, this is expensive when there's a large number of points and segments.

Reply:
Theoretically, yes. I guess it could be optimised further by having a smarter data structure (ideas welcome), but I feel the feature/UI itself would have become unusable from having too many points or segments, before the path actually runs slow.

Reply:
Some of our use cases involve having lots of segments (e.g., annotating speech vs music segments), although in this case we may not be using the events. We'd need to do some measurements to see if optimisation is needed - but if so, I'd expect a binary search approach would be faster, for larger numbers of events.

Reply:
mm, yep, i thought not to optimise it prematurely. I think any slowdown here would be more because of a GC issue, and the worst i can think of is when dragging events are emitted at a very high rate with thousands of segments defined (in which case other parts of the current UI might struggle just the same)